File: etc/ruby/BaseXClient.rb
# Language Binding for BaseX.
# Works with BaseX 6.3.1 and later
# Documentation: http://basex.org/api
#
# (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License

require 'socket'
require 'digest/md5'

# see readme.txt
class Session
  def initialize(host, port, username, pw)

    # create server connection
    @socket = TCPSocket.open(host, port)

    # receive timestamp
    ts = receive
    hash = Digest::MD5.new
    hash.update(hash.hexdigest(pw))
    hash.update(ts)

    # send username and hashed password/timestamp
    send(username)
    send(hash.hexdigest())

    # evaluate success flag
    if read != "\0"
      raise "Access denied."
    end

  end

  # see readme.txt
  def execute(com)
    # send command to server
    send(com)

    # receive result
    result = receive
    @info = receive
    if ok != true
      raise @info
    end
    return result
  end
  
  # see readme.txt
  def query(cmd)
    return Query.new(self, cmd)
  end

  # see readme.txt
  def info()
    return @info
  end

  # see readme.txt
  def close()
    send("exit")
    @socket.close
  end

  # Receives a string from the socket.
  def receive()
    complete = ""
    while ((t = read) != "\0")
    complete += t
    end
    return complete
  end
  
  # Sends the defined str.
  def send(str)
    @socket.write(str + "\0")
  end

  # Returns a single byte from the socket.
  def read()
    return @socket.read(1)
  end
  
  # Returns success check. 
  def ok()
    return read == "\0"
  end
end

class Query
  # see readme.txt
  def initialize(s, q)
    @session = s
    @id = exec("\0", q)
  end
  
  # see readme.txt
  def init()
    return exec("\4", @id)
  end
  
  # see readme.txt
  def bind(name, value)
    exec("\3", @id + "\0" + name + "\0" + value + "\0")
  end
  
  # see readme.txt  
  def more()
    @next = exec("\1", @id)
    return @next.length != 0
  end
  
  # see readme.txt
  def next()
    return @next
  end
  
  # see readme.txt
  def execute()
    return exec("\5", @id)
  end
  
  # see readme.txt
  def info()
    return exec("\6", @id)
  end
  
  # see readme.txt
  def close()
    return exec("\2", @id)
  end
  
  # see readme.txt
  def exec(cmd, arg)
    @session.send(cmd + arg)
    s = @session.receive
    if @session.ok != true
      raise @session.receive
    end
    return s
  end
end

File: etc/python/BaseXClient.py
# Language Binding for BaseX.
# Works with BaseX 6.3.1 and later
# Documentation: http://basex.org/api
#
# (C) BaseX Team 2005-11, ISC License

import hashlib, socket, array
import string

class Session():

    # see readme.txt
    def __init__(self, host, port, user, pw):
        # allocate buffer for speeding up communication
        self.__buf = array.array('B', chr(0) * 0x1000)
        self.init()
        
        # create server connection
        self.__s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.__s.connect((host, port))
        
        # receive timestamp
        ts = self.readString()
        
        # send username and hashed password/timestamp
        m = hashlib.md5()
        m.update(hashlib.md5(pw).hexdigest())
        m.update(ts)
        self.send(user + chr(0) + m.hexdigest())
        
        # evaluate success flag
        if self.__s.recv(1) != chr(0):
            raise IOError('Access Denied.')

    # see readme.txt
    def execute(self, com):
        # send command to server
        self.send(com)
        
        # receive result
        result = self.receive()
        self.__info = self.readString()
        if not self.ok():
            raise IOError(self.__info)
        return result
  
    # see readme.txt
    def query(self, q):
        return Query(self, q)
    
    # see readme.txt
    def add(self, name, target, content):
        self.__s.send(chr(9) + name + chr(0) + target + chr(0) + content + chr(0))
        self.__info = self.readString()
        if not self.ok():
            raise IOError(self.info())

    # see readme.txt
    def info(self):
        return self.__info

    # see readme.txt
    def close(self):
        self.send('exit')
        self.__s.close()

    # Initializes the byte transfer.
    def init(self):
        self.__bpos = 0
        self.__bsize = 0

    # Receives a string from the socket.        
    def readString(self):
        strings = []
        found = False
        while not found:
            found, substr = self.read_until(0)
            strings.append(substr)
        return string.join(strings, "")

    # Returns a single byte from the socket.
    def read(self):
        # Cache next bytes
        if self.__bpos == self.__bsize:
            self.__bsize = self.__s.recv_into(self.__buf)
            self.__bpos = 0
        b = self.__buf[self.__bpos]
        self.__bpos += 1
        return b
    
    # Reads until byte is found.
    def read_until(self, byte):
        # Cache next bytes
        if self.__bpos >= self.__bsize:
            self.__bsize = self.__s.recv_into(self.__buf)
            self.__bpos = 0
            #print self.__buf[:100]
        found = False
        substr = ""
        try:
            pos = self.__buf[self.__bpos:self.__bsize].index(byte)
            found = True
            substr = self.__buf[self.__bpos:pos].tostring()
            self.__bpos = self.__bpos + pos + 1
        except ValueError:
            substr = self.__buf[self.__bpos:self.__bsize].tostring()
            self.__bpos = self.__bsize
        return (found, substr)
    
    # Sends the defined str.
    def send(self, str):
        self.__s.send(str + chr(0))
  
    # Returns success check.
    def ok(self):
        return self.read() == 0
   
    # Returns the received string.
    def receive(self):
        self.init()
        return self.readString()
   
class Query():
    # see readme.txt
    def __init__(self, session, q):
        self.__session = session
        self.__id = self.execu(chr(0), q)
  
    # see readme.txt  
    def init(self):
        return self.execu(chr(4), self.__id)
    
    # see readme.txt  
    def bind(self, name, value):
        self.execu(chr(3), self.__id + chr(0) + name + chr(0) + value + chr(0))
  
    # see readme.txt
    def more(self):
        self.__next = self.execu(chr(1), self.__id)  
        return len(self.__next) != 0  
    
    # see readme.txt
    def next(self):
        return self.__next
  
    # see readme.txt  
    def execute(self):
        return self.execu(chr(5), self.__id)
  
    # see readme.txt  
    def info(self):
        return self.execu(chr(6), self.__id)
  
    # see readme.txt  
    def close(self):
        return self.execu(chr(2), self.__id)
  
    # see readme.txt  
    def execu(self, cmd, arg):
        self.__session.send(cmd + arg)
        s = self.__session.receive()
        if not self.__session.ok():
            raise IOError(self.__session.readString())
        return s